### DISCUSSION TITLE
Tor with Shadow

### DISCUSSION THREAD

**Discussion by suriasarath** [shadow/shadow#3237] [General]
I am new to Shadow. Previously,  I'm currently attempting to run Tor with Shadow, but I'm encountering the following issue. Could you please help me fix it?

**The following error I received from shadow.log.**

```text
00:00:00.011803 [12069:shadow-worker] 00:00:01.000000000 [ERROR] [fileserver:11.0.0.5] [process.rs:1250] [shadow_rs::host::process] process 'fileserver.tor-gencert.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.020717 [12068:shadow-worker] 00:00:01.000000000 [ERROR] [4uthority:11.0.0.1] [process.rs:1250] [shadow_rs::host::process] process '4uthority.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.047016 [12069:shadow-worker] 00:01:00.000000000 [ERROR] [relay3:11.0.0.8] [process.rs:1250] [shadow_rs::host::process] process 'relay3.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.047146 [12070:shadow-worker] 00:01:00.000000000 [ERROR] [relay1:11.0.0.6] [process.rs:1250] [shadow_rs::host::process] process 'relay1.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.047148 [12071:shadow-worker] 00:01:00.000000000 [ERROR] [exit2:11.0.0.4] [process.rs:1250] [shadow_rs::host::process] process 'exit2.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.047293 [12068:shadow-worker] 00:01:00.000000000 [ERROR] [exit1:11.0.0.3] [process.rs:1250] [shadow_rs::host::process] process 'exit1.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.062313 [12069:shadow-worker] 00:01:00.000000000 [ERROR] [relay4:11.0.0.9] [process.rs:1250] [shadow_rs::host::process] process 'relay4.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.062454 [12071:shadow-worker] 00:01:00.000000000 [ERROR] [relay2:11.0.0.7] [process.rs:1250] [shadow_rs::host::process] process 'relay2.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.101879 [12070:shadow-worker] 00:10:00.000000000 [ERROR] [client:11.0.0.2] [process.rs:1250] [shadow_rs::host::process] process 'client.tor-gencert.1000' exited with status Normal(1); expected end state was exited: 0 but was exited: 1
00:00:00.147327 [12071:shadow-worker] 00:15:00.000000000 [ERROR] [torclient:11.0.0.10] [process.rs:1250] [shadow_rs::host::process] process 'torclient.tor.1000' exited with status Normal(1); expected end state was running but was exited: 1
00:00:00.185841 [12071:shadow-worker] 00:25:00.000000000 [ERROR] [torclient:11.0.0.10] [process.rs:1250] [shadow_rs::host::process] process 'torclient.tor-gencert.1001' exited with status Normal(1); expected end state was exited: 0 but was exited: 1
00:00:00.212992 [12052:shadow] n/a [ERROR] [n/a] [main.rs:397] [shadow_rs::core::main::export] Failed to run the simulation
00:00:00.213071 [12052:shadow] n/a [ERROR] [n/a] [main.rs:397] [shadow_rs::core::main::export] 
00:00:00.213146 [12052:shadow] n/a [ERROR] [n/a] [main.rs:397] [shadow_rs::core::main::export] Caused by:
00:00:00.213214 [12052:shadow] n/a [ERROR] [n/a] [main.rs:397] [shadow_rs::core::main::export]     11 managed processes in unexpected final state
```

**This error from console**

```text
** Starting Shadow v3.0.0-1-g2789ee12 2023-05-18--19:09:17 with GLib v2.66.8
** Shadow did not complete successfully: Failed to run the simulation
**   11 managed processes in unexpected final state
** See the log for details
```

**Comment by stevenengler** [comment]
How did you generate the Shadow configuration? With tornettools? Or is this from an example?

The tor and tor-gencert processes are exiting with a non-zero exit code indicating that they had an error. Can you post the stdout/stderr logs for one of the failed processes? For example: `shadow.data/hosts/relay1/tor.1000.stdout` and `shadow.data/hosts/relay1/tor.1000.stderr`.

Can you also post the output of `tor --version`?

**Reply by suriasarath** [reply]
Thanks for the replay.  Tor version is 4.7.13, I am using  shadow  configuration from example.
**shadow.yaml**
general:
  stop_time: 30 min
network:
  graph:
    type: gml
    inline: |
      graph [
        directed 0
        node [
          id 0
          host_bandwidth_down "130 Mbit"
          host_bandwidth_up "130 Mbit"
        ]
        edge [
          source 0
          target 0
          latency "50 ms"
          jitter "0 ms"
          packet_loss 0.0
        ]
      ]
hosts:
  fileserver:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/tools/tor-gencert
      # See https://shadow.github.io/docs/guide/compatibility_notes.html#libopenblas
      environment: { OPENBLAS_NUM_THREADS: "1" }
      args: ../../../conf/tgen.server.graphml.xml
      start_time: 1
      expected_final_state: running
  4uthority:
    network_node_id: 0
    ip_addr: 100.0.0.1
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address 4uthority --Nickname1 4uthority
            --defaults-torrc torrc-defaults -f torrc
      start_time: 1
      expected_final_state: running
  exit1:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address exit1 --Nickname2 exit1
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      expected_final_state: running
  exit2:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address exit2 --Nickname3 exit2
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      expected_final_state: running
  relay1:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address relay1 --Nickname4 relay1
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      expected_final_state: running
  relay2:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address relay2 --Nickname5 relay2
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      expected_final_state: running
  relay3:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address relay3 --Nickname6 relay3
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      expected_final_state: running
  relay4:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address relay4 --Nickname7 relay4
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      expected_final_state: running
  client:
    network_node_id: 0
    processes:
    - path: /home/suria/tor-tor-0.4.7.13/src/tools/tor-gencert
      # See https://shadow.github.io/docs/guide/compatibility_notes.html#libopenblas
      environment: { OPENBLAS_NUM_THREADS: "1" }
      args: ../../../conf/tgen.client.graphml.xml
      start_time: 600
  torclient:
    network_node_id: 0
    processes:
    - path: ~/tor-tor-0.4.7.13/src/app/tor
      args: --Address torclient --Nickname8 torclient
            --defaults-torrc torrc-defaults -f torrc
      start_time: 900
      expected_final_state: running
    - path: /home/suria/tor-tor-0.4.7.13/src/tools/tor-gencert
      # See https://shadow.github.io/docs/guide/compatibility_notes.html#libopenblas
      environment: { OPENBLAS_NUM_THREADS: "1" }
      args: ../../../conf/tgen.torclient.graphml.xml
      start_time: 1500

**Reply by robgjansen** [reply]
If you still want help, we'll need details from the log files. As requested by @stevenengler, you should post the stdout/stderr logs for one of the failed processes. For example, you should post the contents of: 
- `shadow.data/hosts/relay1/tor.1000.stdout` and 
- `shadow.data/hosts/relay1/tor.1000.stderr`.
